gen/type.go

```diff
@@ -157,6 +157,10 @@ func typeName(g Generator, spec compile.TypeSpec) (string, error) {
 		if err != nil {
 			return "", err
 		}
+		if !isPrimitiveType(s.KeySpec) {
```

Comment:
can we create a new function `isHashable` which does this check and is used here and below.